### DISCUSSION TITLE
How can I partially and temporarily mock a local module?

### DISCUSSION THREAD

**Discussion by fregante** [vitest-dev/vitest#2672] [Q&A] [answered]
Let's say I have this module:

```js
// index.js
export function alpha() {
  return 'a'
}
export function beta() {
  return 'b' + alpha();
}
```

In my test file, I want to test each export separately, without overlapping the logic (in reality, `alpha` is more complex than this)

I'd expect to write something similar to this:

```js
// Test the real alpha implementation
test('alpha', () => {
  expect(alpha()).toBe('a')
});

// Then only mock the alpha export
const alphaMock = vi.mocked(alpha)

test('beta', () => {
  alphaMock.mockImplementation(() => 'MOCKED');
  expect(beta()).toBe('bMOCKED')
  expect(alphaMock).toHaveBeenCalled();
});
```

But that of course is wrong. If I add `vi.mock('./index.js')` the first test will also fail.

I can't figure out how to mock `index.js` while simultaneously allowing the real implementation to show up at times.

---

Side note: The easiest solution is probably to split up index.js and/or index.test.js, but I wonder if there's a way that doesn't require that.

**Comment by sheremet-va** [comment, marked as answer]
There is no way to do that in Vitest. Methods should be in separate files.

It is possible to do in Jest with ts-jest, I think? Or babel. One of them has a bug and transforms code incorrectly, allowing Jest to spy on exported method and affect its calls inside single file.

Under usual circumstances in JS it is impossible to intercept that call, unless it's on some kind of an object.

To better understand this, look at the transformed code (I made them into arrow functions):

```js
// jest version
exports.alpha = () => 'a'
exports.beta = () => 'b' + exports.alpha()
```

```js
// vitest version
const alpha = () => 'a'
const beta = () => 'b' + alpha()

exports.aplha = aplha
exports.beta = beta
```

You can see in the second version that `beta` calls `alpha` directly instead of a function on "exports" object.

**Reply by fregante** [reply]
That explains a lot. Thank you for the detailed answer!
